study1/week3/pascals-triangle.java

```diff
@@ -0,0 +1,34 @@
+//https://leetcode.com/problems/pascals-triangle/submissions/
+
+class Solution {
+    public List<List<Integer>> generate(int numRows) {
+        //When numRows is n, time complexity is O(n^n)
+        //Because, the first for loop -> visits n times and the inner for-loop -> visits up to n times 
+        //(because j repeats at most i times, if i is n then j also n)
+        
+        //When numRows is n, space complexity is O(n)
+        //Because, n lists are stored in result list, so the length of the result list is n.
+        //And then in also inner for-loop n lists are stored in semiResult list.
+        //So O(n + n) = O(2n) = O(n)
+      
+        List<List<Integer>> result = new ArrayList<>();
+
+        for (int i = 0; i < numRows; i++) { //until numRow
+            List<Integer> semiResult = new ArrayList<>();
+
+            for (int j = 0; j <= i; j++) { //j <= i -> 1st -> 1 element, 2nd -> 2 elements, 3rd -> 3 elemets, ...
+                if (j == 0) {
```

Comment:
`j == 0` is for the first element and `j == i` is for the last element, so you can move these outside of the loop and reduce the number of comparisons. so doing the following would simplify and reduce comparisons: 
```java
semiResult.add(1);
for (int j = 1; j < i; j++) {
   semiResult.add(result.get(i - 1).get(j - 1) + result.get(i - 1).get(j));
}
semiResult.add(1);
```